text/0039-service-overhaul.md

```diff
@@ -0,0 +1,462 @@
+  - `"stream_key_link"` (optional): link where the user can find his stream key.
+  - `"alt_names"` (optional): alternative service name.
+  - `"servers"`: array of servers.
+    - `"name"`: name of the server.
+    - `"url"`: url of the server.
+  - `"recommended"` (it depends): "recommended" settings that become the default when using this service.
+    - `"output"` (not optional if using a protocol different from RTMP(S)): indicates which output **shall** be used because the service use a protocol different from RTMP(S).
+    - `"keyint"` (optional): default keyframe interval for this service.
+    - `"bframes"` (optional): default b-frames for this service.
+    - `"profile"` (optional): default H264 encoder profile for this service.
+    - `"x264opts"` (optional): default option for x264 for this service.
+    - `"max video bitrate"` (optional): maximum video bitrate for this service.
+    - `"max audio bitrate"` (optional): maximum audio bitrate for this service.
+    - `"max fps"` (optional): maximum frame per second for this service.
+    - `"supported resolutions"` (optional): array of supported resolutions for this service.
+    - `"bitrate matrix"` (optional): array of maximum bitrate for a certain resolution/FPS combo for this service.
+      - `"res"`: resolution.
+      - `"fps"`: frame per seconds.
+      - `"max bitrate"`: maximum video bitrate.
+
+##### Issues with this format
+- The naming scheme is a mix of object name with space and some with underscores.
+- About `"output"`, OBS consider every service as RTMP if not added and it's not a recommended settings at all. It makes OBS use the right protocol for the service. Also this prevent a service of being multi-protocol.
+- About `"recommended"`, most of the options seems to H264 related
+- `"common"`, the name makes it not understandable at the first sight maybe adding some documention would be good thing.
+
+#### New format (WIP)
+The new format will be parsable with a new interface library named `service-json-parser` to allow first-party plugins to read this format rather than recreating the wheel.
+
+The new format will be considered as the version 4.
+
+This format allow some specific additions parsed by getter added to the concerned plugin.
+
+Multi-service plugins should have a JSON root that look like this:
+```json
+{
+    "format_version": 4,
+    "services": []
+}
+```
+`"services"` is an array of service objects.
+
+Service plugin like integrations one should have a JSON root that look like this:
+```json
+{
+    "format_version": 4,
+    "service": {}
+}
+```
+`"service"` is the service object itself.
+
+##### Plugins specific additions
+Those additions are added like extensions to the service object format.
+
+Getters for those customisations are implemented in the plugin itself.
+- In `obs-services`:
+  - `"id"` (required): service identifier, services are no longer identified by their name. Added because this plugin is multi-services.
```

Comment:
I am unsure how best to handle an ID - services sometimes change their name or address, and often no longer want it associated with the old name. This means the plugin author would have to keep that in mind and create an ID that identifies the service without.. naming it. Maybe this should be a UUID-type field that has to be generated?

Reply:
An UUID will maybe make the user unable to detect which service is used through scripting or even websocket. If one day user do want to detect it through plugins or scripts.

I added the use of ID to avoid the creation of `alt_name` form the actual JSON, UUID can also fix the issue but are not "user that may use scripting on it" friendly.